### PR TITLE
bpo-31484: Cache single-character strings outside of the Latin1 range.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue31484

### CODE REVIEW

**Review by methane — APPROVED** [review]
LGTM

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -2038,10 +2110,11 @@ PyUnicode_FromWideChar(const wchar_t *u, Py_ssize_t size)
...
         _Py_RETURN_UNICODE_EMPTY();
 
-    /* Single character Unicode objects in the Latin-1 range are
-       shared when using this constructor */
-    if (size == 1 && (Py_UCS4)*u < 256)
-        return get_latin1_char((unsigned char)*u);
+    /* Single character Unicode objects can be shared when using this
+       constructor. */
+    if (size == 1) {
+        return unicode_char((Py_UCS4)*u);
```
Comment (methane) [inline Objects/unicodeobject.c]:
```
    def test_issue17223(self):
        # this used to crash
        if sizeof_wchar == 4:
            # U+FFFFFFFF is an invalid code point in Unicode 6.0
            invalid_str = b'\xff\xff\xff\xff'
        else:
            # PyUnicode_FromUnicode() cannot fail with 16-bit wchar_t
            self.skipTest("specific to 32-bit wchar_t")
        a = array.array('u', invalid_str)
        self.assertRaises(ValueError, a.tounicode)
        self.assertRaises(ValueError, str, a)
```

PyUnicode_FromWideChar() should raise ValueError when u is out of unicode range.

Comment (serhiy-storchaka) [reply Objects/unicodeobject.c]: Does not it do this? `unicode_char()` is an implementation of Python's `chr()`.

Comment (methane) [reply Objects/unicodeobject.c]: `    assert(ch <= MAX_UNICODE);` is in top of `unicode_char()`

**Review by methane — CHANGES_REQUESTED** [review]

File: Objects/unicodeobject.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Objects/unicodeobject.c]: Good catch!

**Comment by serhiy-storchaka** [comment]
I expect the Japan Inquisition!

**Review by methane — APPROVED** [review]

**Comment by taleinat** [comment]
@serhiy-storchaka, shouldn't this go in? Just add a NEWS entry...

**Comment by serhiy-storchaka** [comment]
I have doubts because of diverse feedback on this proposition. This PR adds complexity in the code, and the gain is small. Most likely I'll close this PR unless other core devs have good need of this microoptimization.

**Comment by csabella** [comment]
@serhiy-storchaka, based on your last comment, do you think this should be closed now?

**Comment by taleinat** [comment]
@serhiy-storchaka, let's close this PR?

**Pull request closed without merging**